# Guided Project: Exploring eBay Car Sales Data

## Introduction

In this project, we will focus on cleaning up the dataset and, once that's done, analyze the used car listings.
The dataset we will work with can be found [here](https://data.world/data-society/used-cars-data). It contains used cars listing data from German eBay and was originally uploaded on [Kaggle](https://www.kaggle.com/) by the user [orgesleka](https://www.kaggle.com/orgesleka) in the already cleaned-up version. Our version of the dataset containes a sample of 50,000 data points from the full dataset and has been dirtied a bit in order to more closely resemble an actual scraped dataset.

## Data Dictionary

* `dateCrawled` - When this ad was first crawled. All field-values are taken from this date.
* `name` - Name of the car.
* `seller` - Whether the seller is private or a dealer.
* `offerType` - The type of listing
* `price` - The price on the ad to sell the car.
* `abtest` - Whether the listing is included in an A/B test.
* `vehicleType` - The vehicle Type.
* `yearOfRegistration` - The year in which the car was first registered.
* `gearbox` - The transmission type.
* `powerPS` - The power of the car in PS.
* `model` - The car model name.
* `kilometer` - How many kilometers the car has driven.
* `monthOfRegistration` - The month in which the car was first registered.
* `fuelType` - What type of fuel the car uses.
* `brand` - The brand of the car.
* `notRepairedDamage` - If the car has a damage which is not yet repaired.
* `dateCreated` - The date on which the eBay listing was created.
* `nrOfPictures` - The number of pictures in the ad.
* `postalCode` - The postal code for the location of the vehicle.
* `lastSeenOnline` - When the crawler saw this ad last online.

## Import pandas and NumPy

In [1]:
import pandas as pd
import numpy as np

In [2]:
autos = pd.read_csv("autos.csv", encoding = 'Latin-1')

Let's explore the dataset:

In [3]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [4]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

In [5]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

From the above, we can observe:
* we have 5 columns with null values - `vehicleType`, `gearbox`, `model`, `fuelType` and `notRepairedDamage` - all of these contain less than 20% of null values
* data is in German, so in order to make it easier for non-German speaking population, we it might be a good idea to replace/translate certain columns to English.
* most of the data is in the form of strings
* column names are written in camelcase, instead of our preferred snakecase

## Renaming columns

We will rename certain columns to a more understandable format, and convert all of the columns from snakecase to camelcase:

In [6]:
autos.rename({"yearOfRegistration" : "registration_year"}, axis = 1, inplace = True)

In [7]:
autos.rename({"monthOfRegistration" : "registration_month"}, axis = 1, inplace = True)

In [8]:
autos.rename({"notRepairedDamage" : "unrepaired_damage"}, axis = 1, inplace = True)

In [9]:
autos.rename({"dateCreated" : "ad_created"}, axis = 1, inplace = True)

In [10]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  registration_year  \
0      privat   Angebot   $5,000  control         bus               2004   
1      privat   Angebot   $8,500  control   limousine               1997   
2      privat   Angebot   $8,990     test   limousine               2009   
3      privat   Angebot   $4,350  control  kleinwagen               2007   
4      privat   Angebot   $1,350     test       kombi               2003   
...       ...       ...      ...      ...         ...                ...   
49995  privat   Angebot  $24,900  control   limousine               2011   
49996  privat   Angebot   $1,980  control      cabrio               1996   
49997  privat   Angebot  $13,200     test      cabrio               2014   
49998  privat   Angebot  $22,900  control       kombi               2013   
49999  privat   Angebot   $1,250  control   limousine               1996   

         gearbox  powerPS   model   odometer  registration_month fuelType  \
0        manuell      158  andere  150,000km                   3      lpg   
1      automatik      286     7er  150,000km                   6   benzin   
2        manuell      102    golf   70,000km                   7   benzin   
3      automatik       71  fortwo   70,000km                   6   benzin   
4        manuell        0   focus  150,000km                   7   benzin   
...          ...      ...     ...        ...                 ...      ...   
49995  automatik      239      q5  100,000km                   1   diesel   
49996    manuell       75   astra  150,000km                   5   benzin   
49997  automatik       69     500    5,000km                  11   benzin   
49998    manuell      150      a3   40,000km                  11   diesel   
49999    manuell      101  vectra  150,000km                   1   benzin   

            brand unrepaired_damage           ad_created  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
49995       82131  2016-04-0

The above way with using `.rename` requires a lot of code, and is suitable for renaming a single column only, so to rename mutiple columns we can use `.columns` instead and simply assign the new column names in the series:

In [11]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'registration_year', 'gearbox', 'powerPS', 'model',
       'odometer', 'registration_month', 'fuelType', 'brand',
       'unrepaired_damage', 'ad_created', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [12]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_photos', 'postal_code',
       'last_seen']
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  num_photos  postal_code  \
0     peugeot              nein  2016-03-26 00:00:00           0        79588   
1         bmw              nein  2016-04-04 00:00:00           0        71034   
2  volkswagen              nein  2016-03-26 00:00:00           0        35394   
3       smart              nein  2016-03-12 00:00:00           0        33729   
4        ford              nein  2016-04-01 00:00:00           0        39218   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37  
3  2016-03-15 03:16:28  
4  2016-04-01 14:38:50

Now, we have column names in more understandable format, and in snakecase.

## Exploring and cleaning columns

In [13]:
autos.describe(include = 'all')

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-22 09:51:06  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  num_photos   postal_code            last_seen  
count                 50000     50000.0  50000.000000                50000  
unique                   76         NaN           NaN                39481  
top     2016-04-03 00:00:00         NaN           NaN  2016-04-07 06:17:27  
freq                   1946         NaN           NaN                    8  
mean                    NaN         0.0  50813.627300                  NaN  
std                     NaN         0.0  25779.747957                  NaN  
min                     NaN         0.0   1067.000000                  NaN  
25%                     NaN         0.0  30451.000000                  NaN  
50%                     NaN         0.0  49577.000000                  NaN  
75%                     NaN         0.0  71540.000000                  NaN  
max                     NaN         0.0  99998.000000                  NaN

Let's do the low hanging fruit first - convert the `price` and `odometer` columns to numerical, and rename columns to include the unit of measure:

In [14]:
autos["price"].head()

0    $5,000
1    $8,500
2    $8,990
3    $4,350
4    $1,350
Name: price, dtype: object

In [15]:
autos["price"] = (autos["price"]
                    .str.replace("$", "")
                    .str.replace(",", "")
                    .astype(int)
                 )
                    
autos["price"].head()

0    5000
1    8500
2    8990
3    4350
4    1350
Name: price, dtype: int64

In [16]:
autos.rename({"price" : "price_usd"}, axis = 1, inplace = True)

In [17]:
autos["odometer"].head()

0    150,000km
1    150,000km
2     70,000km
3     70,000km
4    150,000km
Name: odometer, dtype: object

In [18]:
autos["odometer"] = (autos["odometer"]
                    .str.replace(",", "")
                    .str.replace("km", "")
                    .astype(int)
                 )
                    
autos["odometer"].head()

0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer, dtype: int64

In [19]:
autos.rename({"odometer" : "odometer_km"}, axis = 1, inplace = True)

In [20]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   date_crawled        50000 non-null  object
 1   name                50000 non-null  object
 2   seller              50000 non-null  object
 3   offer_type          50000 non-null  object
 4   price_usd           50000 non-null  int64 
 5   ab_test             50000 non-null  object
 6   vehicle_type        44905 non-null  object
 7   registration_year   50000 non-null  int64 
 8   gearbox             47320 non-null  object
 9   power_ps            50000 non-null  int64 
 10  model               47242 non-null  object
 11  odometer_km         50000 non-null  int64 
 12  registration_month  50000 non-null  int64 
 13  fuel_type           45518 non-null  object
 14  brand               50000 non-null  object
 15  unrepaired_damage   40171 non-null  object
 16  ad_created          50

Now that the low hanging fruit has been taken care of, let's explore column by columnd to determine what else needs to be done. We will list all the issues we observe, and then step by step we can start cleaning the columns.

In [21]:
autos.describe(include = 'all')

date_crawled         name  seller offer_type     price_usd  \
count                 50000        50000   50000      50000  5.000000e+04   
unique                48213        38754       2          2           NaN   
top     2016-03-22 09:51:06  Ford_Fiesta  privat    Angebot           NaN   
freq                      3           78   49999      49999           NaN   
mean                    NaN          NaN     NaN        NaN  9.840044e+03   
std                     NaN          NaN     NaN        NaN  4.811044e+05   
min                     NaN          NaN     NaN        NaN  0.000000e+00   
25%                     NaN          NaN     NaN        NaN  1.100000e+03   
50%                     NaN          NaN     NaN        NaN  2.950000e+03   
75%                     NaN          NaN     NaN        NaN  7.200000e+03   
max                     NaN          NaN     NaN        NaN  1.000000e+08   

       ab_test vehicle_type  registration_year  gearbox      power_ps  model  \
count    50000        44905       50000.000000    47320  50000.000000  47242   
unique       2            8                NaN        2           NaN    245   
top       test    limousine                NaN  manuell           NaN   golf   
freq     25756        12859                NaN    36993           NaN   4024   
mean       NaN          NaN        2005.073280      NaN    116.355920    NaN   
std        NaN          NaN         105.712813      NaN    209.216627    NaN   
min        NaN          NaN        1000.000000      NaN      0.000000    NaN   
25%        NaN          NaN        1999.000000      NaN     70.000000    NaN   
50%        NaN          NaN        2003.000000      NaN    105.000000    NaN   
75%        NaN          NaN        2008.000000      NaN    150.000000    NaN   
max        NaN          NaN        9999.000000      NaN  17700.000000    NaN   

          odometer_km  registration_month fuel_type       brand  \
count    50000.000000        50000.000000     45518       50000   
unique            NaN                 NaN         7          40   
top               NaN                 NaN    benzin  volkswagen   
freq              NaN                 NaN     30107       10687   
mean    125732.700000            5.723360       NaN         NaN   
std      40042.211706            3.711984       NaN         NaN   
min       5000.000000            0.000000       NaN         NaN   
25%     125000.000000            3.000000       NaN         NaN   
50%     150000.000000            6.000000       NaN         NaN   
75%     150000.000000            9.000000       NaN         NaN   
max     150000.000000           12.000000       NaN         NaN   

       unrepaired_damage           ad_created  num_photos   postal_code  \
count              40171                50000     50000.0  50000.000000   
unique                 2                   76         NaN           NaN   
top                 nein  2016-04-03 00:00:00         NaN           NaN   
freq               35232                 1946         NaN           NaN   
mean                 NaN                  NaN         0.0  50813.627300   
std                  NaN                  NaN         0.0  25779.747957   
min                  NaN                  NaN         0.0   1067.000000   
25%                  NaN                  NaN         0.0  30451.000000   
50%                  NaN                  NaN         0.0  49577.000000   
75%                  NaN                  NaN         0.0  71540.000000   
max                  NaN                  NaN         0.0  99998.000000   

                  last_seen  
count                 50000  
unique                39481  
top     2016-04-07 06:17:27  
freq                      8  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

In [22]:
autos["seller"].value_counts()

privat        49999
gewerblich        1
Name: seller, dtype: int64

We can see that the `seller` column contains 49999 identical values. We can drop that column since it is not valuable for our analysis:

In [23]:
autos["offer_type"].value_counts()

Angebot    49999
Gesuch         1
Name: offer_type, dtype: int64

Same goes for the `offer_type` column.
Let's drop those columns using the `df.drop()` method

In [24]:
autos = autos.drop(labels = ["offer_type", "seller"], axis = 1)

We can also observe that the `num_photos``
 column looks weird, so we will explore it:

In [25]:
autos["num_photos"].value_counts()

0    50000
Name: num_photos, dtype: int64

All the values in `num_photos` column are 0, so we can drop that column since it has no value for us.

In [26]:
autos = autos.drop(labels = "num_photos", axis = 1)

In [27]:
autos.columns

Index(['date_crawled', 'name', 'price_usd', 'ab_test', 'vehicle_type',
       'registration_year', 'gearbox', 'power_ps', 'model', 'odometer_km',
       'registration_month', 'fuel_type', 'brand', 'unrepaired_damage',
       'ad_created', 'postal_code', 'last_seen'],
      dtype='object')

In [28]:
autos.shape

(50000, 17)

From the above, we can see we have successfuly dropped 3 unneccessary columns - our new column count is 17

Let's explore some of the numerical columns - `odometer_km` and `price_usd`


In [29]:
autos["odometer_km"].unique().shape

(13,)

In [30]:
autos["odometer_km"].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [31]:
autos["odometer_km"].value_counts().sort_index()

5000        967
10000       264
20000       784
30000       789
40000       819
50000      1027
60000      1164
70000      1230
80000      1436
90000      1757
100000     2169
125000     5170
150000    32424
Name: odometer_km, dtype: int64

When we explore the odometer_km, we can see that the values are rounded and equally incremented, which suggest users had to choose from pre-defined values. We can observe that approx. 65% of vehicles have high mileage (150,000 km), and approx. 80% have mileage of 100,000 km or higher.

In [32]:
autos["price_usd"].unique().shape

(2357,)

In [33]:
autos["price_usd"].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price_usd, dtype: float64

In [34]:
autos["price_usd"].value_counts().sort_index()

0           1421
1            156
2              3
3              1
5              2
            ... 
10000000       1
11111111       2
12345678       3
27322222       1
99999999       1
Name: price_usd, Length: 2357, dtype: int64

There are 1421 rows which a have a price of 0. Those columns are outliers, since definitely no one is giving away their vehicles for free, and since 1421 rows is less than 1% of the total, we can consider dropping those rows

In [35]:
autos["price_usd"].value_counts().sort_index(ascending = False).head(20)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: price_usd, dtype: int64

We can see that there are some unusually high prices - there are 15 vehicles listed with prices above or close to a million, we can probably drop those rows and retain everything priced at 350,000 and less, since that is much more realistic.

In [36]:
autos["price_usd"].value_counts().sort_index(ascending = True).head(50)

0      1421
1       156
2         3
3         1
5         2
8         1
9         1
10        7
11        2
12        3
13        2
14        1
15        2
17        3
18        1
20        4
25        5
29        1
30        7
35        1
40        6
45        4
47        1
49        4
50       49
55        2
59        1
60        9
65        5
66        1
70       10
75        5
79        1
80       15
89        1
90        5
99       19
100     134
110       3
111       2
115       2
117       1
120      39
122       1
125       8
129       1
130      15
135       1
139       1
140       9
Name: price_usd, dtype: int64

When we look at the ascending price order, we can see that there are values which incrementaly increase from 1 USD. SInce ebay is an auction site, it might be possible that these sellers have started the auctions with low values. Since that assumption is viable, we will retain those rows, and only drop rows with prices above 350,000.

In [37]:
autos = autos[autos["price_usd"].between(1, 350100)]
autos["price_usd"].describe()

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price_usd, dtype: float64

We can see that all the outliers have been successfully removed and our prices now range from 1 to 350,000 USD.

In [38]:
autos["power_ps"].unique().shape

(444,)

In [39]:
autos["power_ps"].value_counts().sort_index().head(30)

0     4986
1        5
2        2
3        3
4        4
5       13
6        3
8        2
9        1
10       2
11       4
12       1
14       1
15       5
16       1
18       6
19       3
20       4
21       1
23       3
24       1
25       2
26      34
27       5
29       4
30       3
33       9
34      27
35       2
37       7
Name: power_ps, dtype: int64

In order to better understand what we are looking at, we need to investigate what kind of a unit of measure is PS. A quick Google search tells us:
PS stands for PferdStarke (literally, 'horse strength' in German). This is basically metric horsepower as opposed to the imperial or mechanical measure of horsepower denoted by HP or BHP. 1 PS, equals 4,500 kilogram-metres per minute, which rounds to 32,550 foot-pounds per minute, or 0.9863 of a horsepower.

We can observe that there are 4986 vehicles with horsepower of 0, which is not possible. Maybe sellers had no info of the HP, which is unlikely since it is always specified in the vehicle permit. We can consider removing those rows. We can also consider to remove all rows with horsepower less than 50, since it is highly unlikely that cars have such low HP.

In [40]:
(autos["power_ps"] < 50).sum() / autos.shape[0]

0.11841861422835376

In [41]:
autos["power_ps"].value_counts(normalize = True).head(10)

0      0.102667
75     0.063173
60     0.044023
150    0.040667
140    0.038443
101    0.035396
90     0.034943
116    0.033399
170    0.030186
105    0.028683
Name: power_ps, dtype: float64

In [42]:
autos["power_ps"].value_counts(normalize = True).tail(10)

262      0.000021
571      0.000021
454      0.000021
678      0.000021
430      0.000021
1779     0.000021
1011     0.000021
6512     0.000021
923      0.000021
16312    0.000021
Name: power_ps, dtype: float64

Since cars with less than 50 HP constitute 11% of our data, maybe we can retain those rows, since what we are analysing is sales data, rather than performance data. The distribution looks fine - majority of the vehicles have realistic HP values for European market. Again, we see there is a lot of HP values of zero, which can really point to the fact sellers either do not know the exact HP, or simply haven't bothered to enter it.

## Exploring & cleaning up registration data

In [43]:
autos["registration_year"].unique().shape

(95,)

In [44]:
autos["registration_year"].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

Minimum registration year is 1000, which is not possible. Maximum registration year is 9999 which is also not possible.

In [45]:
autos["registration_year"].value_counts().sort_index(ascending = False).head(20)

9999       3
9000       1
8888       1
6200       1
5911       1
5000       4
4800       1
4500       1
4100       1
2800       1
2019       2
2018     470
2017    1392
2016    1220
2015     392
2014     663
2013     803
2012    1310
2011    1623
2010    1589
Name: registration_year, dtype: int64

In [46]:
autos.describe(include = "all")

date_crawled         name      price_usd ab_test vehicle_type  \
count                 48565        48565   48565.000000   48565        43979   
unique                46882        37470            NaN       2            8   
top     2016-03-29 23:42:13  Ford_Fiesta            NaN    test    limousine   
freq                      3           76            NaN   25019        12598   
mean                    NaN          NaN    5888.935591     NaN          NaN   
std                     NaN          NaN    9059.854754     NaN          NaN   
min                     NaN          NaN       1.000000     NaN          NaN   
25%                     NaN          NaN    1200.000000     NaN          NaN   
50%                     NaN          NaN    3000.000000     NaN          NaN   
75%                     NaN          NaN    7490.000000     NaN          NaN   
max                     NaN          NaN  350000.000000     NaN          NaN   

        registration_year  gearbox      power_ps  model    odometer_km  \
count        48565.000000    46222  48565.000000  46107   48565.000000   
unique                NaN        2           NaN    245            NaN   
top                   NaN  manuell           NaN   golf            NaN   
freq                  NaN    36102           NaN   3900            NaN   
mean          2004.755421      NaN    117.197158    NaN  125770.101925   
std             88.643887      NaN    200.649618    NaN   39788.636804   
min           1000.000000      NaN      0.000000    NaN    5000.000000   
25%           1999.000000      NaN     71.000000    NaN  125000.000000   
50%           2004.000000      NaN    107.000000    NaN  150000.000000   
75%           2008.000000      NaN    150.000000    NaN  150000.000000   
max           9999.000000      NaN  17700.000000    NaN  150000.000000   

        registration_month fuel_type       brand unrepaired_damage  \
count         48565.000000     44535       48565             39464   
unique                 NaN         7          40                 2   
top                    NaN    benzin  volkswagen              nein   
freq                   NaN     29368       10336             34775   
mean              5.782251       NaN         NaN               NaN   
std               3.685595       NaN         NaN               NaN   
min               0.000000       NaN         NaN               NaN   
25%               3.000000       NaN         NaN               NaN   
50%               6.000000       NaN         NaN               NaN   
75%               9.000000       NaN         NaN               NaN   
max              12.000000       NaN         NaN               NaN   

                 ad_created   postal_code            last_seen  
count                 48565  48565.000000                48565  
unique                   76           NaN                38474  
top     2016-04-03 00:00:00           NaN  2016-04-07 06:17:27  
freq                   1887           NaN                    8  
mean                    NaN  50975.745207                  NaN  
std                     NaN  25746.968398                  NaN  
min                     NaN   1067.000000                  NaN  
25%                     NaN  30657.000000                  NaN  
50%                     NaN  49716.000000                  NaN  
75%                     NaN  71665.000000                  NaN  
max                     NaN  99998.000000                  NaN

In [47]:
autos["last_seen"].max()

'2016-04-07 14:58:50'

In [48]:
autos["date_crawled"].min()

'2016-03-05 14:06:30'

Since the max value of last_seen is in 2016, registration years 2017 and above are not valid. In addition, we need to combine year and month of registration to not exceed March 2016, since that is the date ads have been crawled. Any rows with registration months/years newer than February 2016 should be removed.

In [49]:
autos["registration_year"].value_counts().sort_index(ascending = True).head(20)

1000    1
1001    1
1111    1
1800    2
1910    5
1927    1
1929    1
1931    1
1934    2
1937    4
1938    1
1939    1
1941    2
1943    1
1948    1
1950    3
1951    2
1952    1
1953    1
1954    2
Name: registration_year, dtype: int64

For the lower limit of our registration interval, we can see that there are a couple of rows showing registration year from the beginning of the 20th century. Maybe those are old-timers, so before we remove these rows entirely, it is worth to explore further.

In [50]:
autos[autos["registration_year"] < 1911]

date_crawled                                   name  price_usd  \
3679   2016-04-04 00:36:17                             Suche_Auto          1   
10556  2016-04-01 06:02:10                             UNFAL_Auto        450   
22316  2016-03-29 16:56:41  VW_Kaefer.__Zwei_zum_Preis_von_einem.       1500   
22659  2016-03-14 08:51:18                           Opel_Corsa_B        500   
24511  2016-03-17 19:45:11            Trabant__wartburg__Ostalgie        490   
28693  2016-03-22 17:48:41                         Renault_Twingo        599   
30781  2016-03-25 13:47:46       Opel_Calibra_V6_DTM_Bausatz_1:24         30   
32585  2016-04-02 16:56:39                             UNFAL_Auto        450   
45157  2016-03-11 22:37:01                             Motorhaube         15   
49283  2016-03-15 18:38:53                             Citroen_HY       7750   

       ab_test vehicle_type  registration_year  gearbox  power_ps    model  \
3679      test          NaN               1910      NaN         0      NaN   
10556  control          NaN               1800      NaN      1800      NaN   
22316  control          NaN               1000  manuell         0   kaefer   
22659     test          NaN               1910      NaN         0    corsa   
24511  control          NaN               1111      NaN         0      NaN   
28693  control   kleinwagen               1910  manuell         0      NaN   
30781     test          NaN               1910      NaN         0  calibra   
32585  control          NaN               1800      NaN      1800      NaN   
45157  control          NaN               1910      NaN         0      NaN   
49283  control          NaN               1001      NaN         0   andere   

       odometer_km  registration_month fuel_type           brand  \
3679          5000                   0       NaN  sonstige_autos   
10556         5000                   2       NaN      mitsubishi   
22316         5000                   0    benzin      volkswagen   
22659       150000                   0       NaN            opel   
24511         5000                   0       NaN         trabant   
28693         5000                   0    benzin         renault   
30781       100000                   0       NaN            opel   
32585         5000                   2       NaN      mitsubishi   
45157         5000                   0       NaN         trabant   
49283         5000                   0       NaN         citroen   

      unrepaired_damage           ad_created  postal_code            last_seen  
3679                NaN  2016-04-04 00:00:00        40239  2016-04-04 07:49:15  
10556              nein  2016-04-01 00:00:00        63322  2016-04-01 09:42:30  
22316               NaN  2016-03-29 00:00:00        48324  2016-03-31 10:15:28  
22659               NaN  2016-03-14 00:00:00        52393  2016-04-03 07:53:55  
24511               NaN  2016-03-17 00:00:00        16818  2016-04-07 07:17:29  
28693               NaN  2016-03-22 00:00:00        70376  2016-04-06 09:16:59  
30781               NaN  2016-03-25 00:00:00        47638  2016-03-26 23:46:29  
32585              nein  2016-04-02 00:00:00        63322  2016-04-04 14:46:21  
45157               NaN  2016-03-11 00:00:00        90491  2016-03-25 11:18:57  
49283               NaN  2016-03-15 00:00:00        66706  2016-04-06 18:47:20

We can see that the results are mixed - there are some oldtimers, but also there are some cars which clearly did not exist in that time period, e.g. Opel Corsa or Renault Twingo. To ensure better data quality, we can remove all rows with registration years prior to 1911.

Let's check how many rows we have with dates newer than February 2016:

In [51]:
autos[(autos["registration_year"] == 2016) & (autos["registration_month"] > 2)]

date_crawled                                               name  \
135    2016-03-12 11:00:10  Opel_Meriva_B_Panoramadach__Sitz__und_Lenkradh...   
256    2016-04-03 20:50:38                     Passat_1.9TDI_4Motion_Highline   
295    2016-03-28 03:36:22                                     Privat_anbiter   
307    2016-03-15 22:50:48                                       Giessen_ford   
437    2016-03-25 16:39:01                      Mazda__klima_leder__Alufelgen   
...                    ...                                                ...   
49547  2016-03-30 16:49:46                    Smart_Passion_mit_Panorama_Dach   
49852  2016-04-01 04:02:25                                   TOP__Golf_3_1.8l   
49876  2016-03-22 17:57:24                             Audi_a5_3.0_tdi_s_line   
49919  2016-03-10 09:49:43                                         Fiat_Punto   
49938  2016-03-28 18:45:06                     Mercedes_Benz_A_160_Avantgarde   

       price_usd  ab_test vehicle_type  registration_year    gearbox  \
135         8500  control          NaN               2016    manuell   
256         4250     test          NaN               2016    manuell   
295         1000  control          NaN               2016    manuell   
307         2800     test          NaN               2016    manuell   
437          550  control          NaN               2016    manuell   
...          ...      ...          ...                ...        ...   
49547       3600     test          NaN               2016        NaN   
49852       1450  control          NaN               2016        NaN   
49876      14700  control          NaN               2016    manuell   
49919        180     test          NaN               2016    manuell   
49938       2300  control          NaN               2016  automatik   

       power_ps     model  odometer_km  registration_month fuel_type  \
135          81    meriva        90000                   8    diesel   
256         131    passat       150000                  11       NaN   
295           0     astra       150000                   8    benzin   
307         109     focus       150000                   8       NaN   
437           0       NaN       150000                   7    benzin   
...         ...       ...          ...                 ...       ...   
49547        71    fortwo       125000                  10    benzin   
49852        75      golf       150000                   7    benzin   
49876         0       NaN       150000                  10    diesel   
49919        86     punto       150000                   8       NaN   
49938       102  a_klasse       150000                   6    benzin   

               brand unrepaired_damage           ad_created  postal_code  \
135             opel               NaN  2016-03-12 00:00:00        48147   
256       volkswagen               NaN  2016-04-03 00:00:00        31224   
295             opel               NaN  2016-03-28 00:00:00        44147   
307             ford               NaN  2016-03-15 00:00:00        35396   
437            mazda                ja  2016-03-25 00:00:00        70372   
...              ...               ...                  ...          ...   
49547          smart               NaN  2016-03-30 00:00:00        49413   
49852     volkswagen               NaN  2016-04-01 00:00:00        67661   
49876           audi               NaN  2016-03-22 00:00:00        44227   
49919           fiat                ja  2016-03-10 00:00:00        59558   
49938  mercedes_benz              nein  2016-03-28 00:00:00        13507   

                 last_seen  
135    2016-03-22 14:49:31  
256    2016-04-03 20:50:38  
295    2016-03-29 23:18:11  
307    2016-04-07 05:16:07  
437    2016-04-06 20:48:19  
...                    ...  
49547  2016-03-30 16:49:46  
49852  2016-04-05 03:17:29  
49876  2016-03-22 18:40:53  
49919  2016-03-10 10:39:58  
49938  2016-04-07 00:44:35  

[795 rows x 17 columns]

Let's see what is the ratio of these outliers:

In [52]:
((autos["registration_year"] == 2016) & (autos["registration_month"] > 2)).sum() / autos.shape[0]

0.016369813651806855

In [53]:
(autos["registration_year"] < 1911).sum() / autos.shape[0]


0.00020590960568310512

Above we have summed up all the rows which include registration years older than 1911, and all the rows which include registration dates newer than February 2016. Together they add up to approx. 1.6% of our data, so we can go ahead and remove them.

In [54]:
autos = autos[autos["registration_year"] > 1911]
autos = autos[autos["registration_year"] < 2016]
autos = autos[~((autos["registration_year"] == 2016) & (autos["registration_month"] > 2))]
autos["registration_year"].value_counts(normalize = True).head(15)

2000    0.069430
2005    0.064590
1999    0.063732
2004    0.059464
2003    0.059376
2006    0.058738
2001    0.057990
2002    0.054690
1998    0.051984
2007    0.050092
2008    0.048728
2009    0.045869
1997    0.042921
2011    0.035705
2010    0.034957
Name: registration_year, dtype: float64

In [55]:
autos.shape

(45456, 17)

We have reduced our dataset to 45456 rows of data. Registration year distribution looks good, with majority of the data falling into the 1997+ year range.

## Exploring price by brand

First, we will take a look at all the brands in the dataset and select the top brands by percentage.

In [56]:
autos["brand"].unique().shape

(40,)

We can see there are 40 unique brands in the dataset. Let's see which those are:

In [57]:
autos["brand"].value_counts(normalize = True)

volkswagen        0.210533
bmw               0.111184
opel              0.106191
mercedes_benz     0.097347
audi              0.087249
ford              0.069716
renault           0.046551
peugeot           0.029611
fiat              0.025365
seat              0.018105
skoda             0.016631
nissan            0.015246
mazda             0.015136
smart             0.014146
citroen           0.014014
toyota            0.012760
hyundai           0.010076
sonstige_autos    0.009856
volvo             0.009262
mini              0.008778
mitsubishi        0.008184
honda             0.007832
kia               0.007194
alfa_romeo        0.006666
porsche           0.006182
suzuki            0.005918
chevrolet         0.005764
chrysler          0.003542
dacia             0.002662
daihatsu          0.002530
jeep              0.002288
subaru            0.002156
land_rover        0.002112
saab              0.001672
jaguar            0.001606
daewoo            0.001518
trabant           0.001408
r

As expected, the majority of the brands that are offered are European (over 75%), with German brands taking the lead.

We will take the top 80% of the brands for our price analysis: 
* volkswagen
* bmw
* opel
* mercedes_benz
* audi
* ford
* renault
* peugeot
* fiat
* seat
* skoda

Creating an empty dictionary to hold the price data:

In [58]:
brand_mean_prices = {}

We will assign our normalized value count to a new variable, and then use the .index attribute to access the top 80% market share brands:

In [59]:
brands_value_counts = autos["brand"].value_counts(normalize = True)
brands = brands_value_counts[:11].index
brands

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat', 'skoda'],
      dtype='object')

We will now loop over each brand and calculate the mean price. Then we will assign the brand as a key to the dictionary, and calculated mean price for each brand as a value to the key (as integer for better readability).

In [60]:
for b in brands:
    selected_rows = autos[autos["brand"] == b]
    mean_price = selected_rows["price_usd"].mean()
    brand_mean_prices[b] = int(mean_price)

In [61]:
brand_mean_prices

{'volkswagen': 5480,
 'bmw': 8398,
 'opel': 3014,
 'mercedes_benz': 8711,
 'audi': 9349,
 'ford': 3817,
 'renault': 2492,
 'peugeot': 3130,
 'fiat': 2861,
 'seat': 4423,
 'skoda': 6349}

We can see that the #1 brand, Volkswagen, has a very attractive mean price - it is much cheaper than BMW, Mercedes or Audi, while more expensive on average than Opel, Renault or Peugeot. The attractive price and German origin most likely make it so popular.

On the other hand, BMW, Mercedes and Audi are most expensive, but still rank amongst the Top 5.

Opel, Ford, Peugeot and Renault are less expensive than all the above mentioned brands, so it is realistic they will have a large portion of the market share.

## Calculating the mean mileage

Using the same principle as above, we will calculate the mean mileage for each of our selected brands:

In [62]:
brand_mean_mileage = {}

In [63]:
for b in brands:
    selected_rows = autos[autos["brand"] == b]
    mean_mileage = selected_rows["odometer_km"].mean()
    brand_mean_mileage[b] = int(mean_mileage)
    
brand_mean_mileage

{'volkswagen': 128442,
 'bmw': 132462,
 'opel': 129233,
 'mercedes_benz': 130615,
 'audi': 129189,
 'ford': 124001,
 'renault': 127996,
 'peugeot': 127113,
 'fiat': 116647,
 'seat': 120972,
 'skoda': 111005}

## Create a new dataframe for comparison of price vs mileage

We will first use pandas series constructor to convert both `brand_mean_prices` and `brand_mean_mileage` dictionaries to series objects:

In [64]:
bmp_series = pd.Series(brand_mean_prices).sort_values(ascending = False)

In [65]:
bmp_series

audi             9349
mercedes_benz    8711
bmw              8398
skoda            6349
volkswagen       5480
seat             4423
ford             3817
peugeot          3130
opel             3014
fiat             2861
renault          2492
dtype: int64

In [66]:
bmm_series = pd.Series(brand_mean_mileage).sort_values(ascending = False)

In [67]:
bmm_series

bmw              132462
mercedes_benz    130615
opel             129233
audi             129189
volkswagen       128442
renault          127996
peugeot          127113
ford             124001
seat             120972
fiat             116647
skoda            111005
dtype: int64

In [68]:
mean_price_df = pd.DataFrame(bmp_series, columns = ["mean_price"])

In [69]:
mean_price_df

mean_price
audi                 9349
mercedes_benz        8711
bmw                  8398
skoda                6349
volkswagen           5480
seat                 4423
ford                 3817
peugeot              3130
opel                 3014
fiat                 2861
renault              2492

In [70]:
mean_mileage_df = pd.DataFrame(bmm_series, columns = ["mean_mileage"])

In [71]:
mean_mileage_df

mean_mileage
bmw                  132462
mercedes_benz        130615
opel                 129233
audi                 129189
volkswagen           128442
renault              127996
peugeot              127113
ford                 124001
seat                 120972
fiat                 116647
skoda                111005

In [72]:
mean_mileage_df["mean_price"] = mean_price_df

In [73]:
mean_mileage_df

mean_mileage  mean_price
bmw                  132462        8398
mercedes_benz        130615        8711
opel                 129233        3014
audi                 129189        9349
volkswagen           128442        5480
renault              127996        2492
peugeot              127113        3130
ford                 124001        3817
seat                 120972        4423
fiat                 116647        2861
skoda                111005        6349

In [74]:
brand_info = mean_mileage_df

In [75]:
brand_info

mean_mileage  mean_price
bmw                  132462        8398
mercedes_benz        130615        8711
opel                 129233        3014
audi                 129189        9349
volkswagen           128442        5480
renault              127996        2492
peugeot              127113        3130
ford                 124001        3817
seat                 120972        4423
fiat                 116647        2861
skoda                111005        6349

We have merged both series into one dataframe called `brand_info` with values sorted in a descending order. Now we can easily compare prices vs mileage.

We cannot observe a large gap in mileage, but rather a trend that more expensive brands tend to have slightly higher mileage than less expensive brands. Exception is Skoda, which has quite low mileage for the mean price.

Since Mercedes, BMW and Audi mostly make limousines, it may be the reason why these brands have higher mean mileages - limousines are mostly used for long range travel, while cheaper vehicles will mostly be used within the city limits, for commuting.

## Translating German to English

In [76]:
autos.describe(include = "all")

date_crawled      name      price_usd ab_test vehicle_type  \
count                 45456     45456   45456.000000   45456        43928   
unique                43990     34756            NaN       2            8   
top     2016-03-23 18:39:34  BMW_316i            NaN    test    limousine   
freq                      3        75            NaN   23437        12588   
mean                    NaN       NaN    6032.640202     NaN          NaN   
std                     NaN       NaN    8909.181759     NaN          NaN   
min                     NaN       NaN       1.000000     NaN          NaN   
25%                     NaN       NaN    1299.000000     NaN          NaN   
50%                     NaN       NaN    3200.000000     NaN          NaN   
75%                     NaN       NaN    7650.000000     NaN          NaN   
max                     NaN       NaN  345000.000000     NaN          NaN   

        registration_year  gearbox      power_ps  model    odometer_km  \
count        45456.000000    43510  45456.000000  43405   45456.000000   
unique                NaN        2           NaN    243            NaN   
top                   NaN  manuell           NaN   golf            NaN   
freq                  NaN    33798           NaN   3583            NaN   
mean          2002.569672      NaN    118.972963    NaN  125447.465681   
std              6.881330      NaN    186.727578    NaN   39802.490194   
min           1927.000000      NaN      0.000000    NaN    5000.000000   
25%           1999.000000      NaN     75.000000    NaN  100000.000000   
50%           2003.000000      NaN    109.000000    NaN  150000.000000   
75%           2007.000000      NaN    150.000000    NaN  150000.000000   
max           2015.000000      NaN  17700.000000    NaN  150000.000000   

        registration_month fuel_type       brand unrepaired_damage  \
count         45456.000000     42625       45456             37710   
unique                 NaN         7          40                 2   
top                    NaN    benzin  volkswagen              nein   
freq                   NaN     27961        9570             33325   
mean              5.854387       NaN         NaN               NaN   
std               3.657980       NaN         NaN               NaN   
min               0.000000       NaN         NaN               NaN   
25%               3.000000       NaN         NaN               NaN   
50%               6.000000       NaN         NaN               NaN   
75%               9.000000       NaN         NaN               NaN   
max              12.000000       NaN         NaN               NaN   

                 ad_created   postal_code            last_seen  
count                 45456  45456.000000                45456  
unique                   74           NaN                36241  
top     2016-04-03 00:00:00           NaN  2016-04-07 06:17:27  
freq                   1774           NaN                    8  
mean                    NaN  51166.267643                  NaN  
std                     NaN  25746.241556                  NaN  
min                     NaN   1067.000000                  NaN  
25%                     NaN  30855.000000                  NaN  
50%                     NaN  50129.000000                  NaN  
75%                     NaN  72108.000000                  NaN  
max                     NaN  99998.000000                  NaN

In [77]:
autos["vehicle_type"].value_counts()

limousine     12588
kleinwagen    10568
kombi          8923
bus            4028
cabrio         3014
coupe          2459
suv            1958
andere          390
Name: vehicle_type, dtype: int64

In [78]:
translation = ({"kleinwagen" : "small_car", 
                "kombi" : "van", 
                "andere" : "other",
               "limousine" : "limousine",
               "bus" : "bus",
               "cabrio" : "convertible",
               "coupe" : "coupe",
               "suv" : "suv"})
autos["vehicle_type"] = autos["vehicle_type"].map(translation)
autos["vehicle_type"].value_counts()

limousine      12588
small_car      10568
van             8923
bus             4028
convertible     3014
coupe           2459
suv             1958
other            390
Name: vehicle_type, dtype: int64

In [79]:
autos["fuel_type"].value_counts()

benzin     27961
diesel     13887
lpg          641
cng           70
hybrid        36
elektro       16
andere        14
Name: fuel_type, dtype: int64

In [80]:
translation = ({"benzin" : "petrol", 
                "diesel" : "diesel", 
                "lpg" : "lpg",
               "cng" : "cng",
               "hybrid" : "hybrid",
               "elektro" : "electric",
               "andere" : "other"})
autos["fuel_type"] = autos["fuel_type"].map(translation)
autos["fuel_type"].value_counts()

petrol      27961
diesel      13887
lpg           641
cng            70
hybrid         36
electric       16
other          14
Name: fuel_type, dtype: int64

In [81]:
autos["gearbox"].value_counts()

manuell      33798
automatik     9712
Name: gearbox, dtype: int64

In [82]:
translation = ({"manuell" : "manual", 
                "automatik" : "automatic"})
autos["gearbox"] = autos["gearbox"].map(translation)
autos["gearbox"].value_counts()

manual       33798
automatic     9712
Name: gearbox, dtype: int64

In [83]:
autos["unrepaired_damage"].value_counts()

nein    33325
ja       4385
Name: unrepaired_damage, dtype: int64

In [84]:
translation = ({"nein" : "no", 
                "ja" : "yes"})
autos["unrepaired_damage"] = autos["unrepaired_damage"].map(translation)
autos["unrepaired_damage"].value_counts()

no     33325
yes     4385
Name: unrepaired_damage, dtype: int64

## Mileage groups

Since we have concluded mileages are rounded up, they can easily be divided into categories for further analysis. We assume vehicles with lower mileage will have lower mean prices, so let's see if that is correct.

We will start by taking a look once again at odometer value counts:

In [89]:
autos["odometer_km"].value_counts().sort_index()

5000        706
10000       239
20000       735
30000       755
40000       790
50000       988
60000      1116
70000      1173
80000      1354
90000      1647
100000     2024
125000     4772
150000    29157
Name: odometer_km, dtype: int64

We can see that there are 13 mileage categories, we will narrow them down into 3 groups - low, medium and high:

In [95]:
low_mileage = autos[autos["odometer_km"] <= 50000]
medium_mileage = autos[autos["odometer_km"].between(60000, 90000)]
high_mileage = autos[autos["odometer_km"] >= 100000]


In [96]:
print(low_mileage["price_usd"].mean())
print(medium_mileage["price_usd"].mean())
print(high_mileage["price_usd"].mean())

14685.632803228104
10225.208128544424
4401.795983645315


From the above, we can see that our assumption was correct - mean prices drop significantly with the mileage.

Now, let's see which brands are the most prevalent in these mileage categories:

In [104]:
low_mileage["brand"].value_counts(normalize = True).sort_values(ascending = False).head(10)

volkswagen        0.193211
audi              0.084975
opel              0.083314
mercedes_benz     0.074769
bmw               0.070496
ford              0.067648
renault           0.036079
sonstige_autos    0.034417
fiat              0.032044
smart             0.031094
Name: brand, dtype: float64

In [105]:
medium_mileage["brand"].value_counts(normalize = True).sort_values(ascending = False).head(10)

volkswagen       0.176938
opel             0.083743
bmw              0.083554
ford             0.080529
mercedes_benz    0.078450
audi             0.064461
renault          0.041966
fiat             0.035350
peugeot          0.034783
smart            0.032703
Name: brand, dtype: float64

In [106]:
high_mileage["brand"].value_counts(normalize = True).sort_values(ascending = False).head(10)

volkswagen       0.217506
bmw              0.120018
opel             0.112174
mercedes_benz    0.102773
audi             0.090869
ford             0.068367
renault          0.048452
peugeot          0.029984
fiat             0.023114
seat             0.017161
Name: brand, dtype: float64

This is all in line with our previous findings - Volkswagen dominates the market, with Opel, BMW, Mercedes and Audi following.

In [108]:
brands_low_vc = low_mileage["brand"].value_counts(normalize = True).sort_values(ascending = False).head(10)
brands_low = brands_low_vc.index
brands_low

Index(['volkswagen', 'audi', 'opel', 'mercedes_benz', 'bmw', 'ford', 'renault',
       'sonstige_autos', 'fiat', 'smart'],
      dtype='object')

In [109]:
low_mileage_price = {}

for b in brands_low:
    selected_rows = low_mileage[low_mileage["brand"] == b]
    mean_price = selected_rows["price_usd"].mean()
    low_mileage_price[b] = int(mean_price)

In [112]:
lmp_series = pd.Series(low_mileage_price).sort_values(ascending = False)

In [113]:
lmp_series

audi              24618
mercedes_benz     22143
bmw               21304
sonstige_autos    15821
volkswagen        13458
ford              10942
opel               8046
renault            7389
fiat               6660
smart              6022
dtype: int64

In [114]:
brands_medium_vc = medium_mileage["brand"].value_counts(normalize = True).sort_values(ascending = False).head(10)
brands_medium = brands_medium_vc.index
brands_medium

Index(['volkswagen', 'opel', 'bmw', 'ford', 'mercedes_benz', 'audi', 'renault',
       'fiat', 'peugeot', 'smart'],
      dtype='object')

In [115]:
medium_mileage_price = {}

for b in brands_medium:
    selected_rows = medium_mileage[medium_mileage["brand"] == b]
    mean_price = selected_rows["price_usd"].mean()
    medium_mileage_price[b] = int(mean_price)

In [117]:
mmp_series = pd.Series(medium_mileage_price).sort_values(ascending = False)

In [118]:
mmp_series

audi             18485
mercedes_benz    17445
bmw              16872
volkswagen       10788
ford              5951
opel              5853
peugeot           5383
renault           4686
fiat              4473
smart             4220
dtype: int64

In [119]:
brands_high_vc = high_mileage["brand"].value_counts(normalize = True).sort_values(ascending = False).head(10)
brands_high = brands_high_vc.index
brands_high

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat'],
      dtype='object')

In [120]:
high_mileage_price = {}

for b in brands_high:
    selected_rows = high_mileage[high_mileage["brand"] == b]
    mean_price = selected_rows["price_usd"].mean()
    high_mileage_price[b] = int(mean_price)

In [121]:
hmp_series = pd.Series(high_mileage_price).sort_values(ascending = False)

In [122]:
hmp_series

audi             6722
bmw              6642
mercedes_benz    6585
volkswagen       4015
seat             2888
ford             2621
peugeot          2344
opel             2264
fiat             1881
renault          1785
dtype: int64

In [126]:
mileage_group_info = pd.DataFrame(lmp_series, columns = ["mean_price_low_mileage"])

In [128]:
mileage_group_info["mean_price_medium_mileage"] = mmp_series

In [130]:
mileage_group_info["mean_price_high_mileage"] = hmp_series

In [131]:
mileage_group_info

mean_price_low_mileage  mean_price_medium_mileage  \
audi                             24618                    18485.0   
mercedes_benz                    22143                    17445.0   
bmw                              21304                    16872.0   
sonstige_autos                   15821                        NaN   
volkswagen                       13458                    10788.0   
ford                             10942                     5951.0   
opel                              8046                     5853.0   
renault                           7389                     4686.0   
fiat                              6660                     4473.0   
smart                             6022                     4220.0   

                mean_price_high_mileage  
audi                             6722.0  
mercedes_benz                    6585.0  
bmw                              6642.0  
sonstige_autos                      NaN  
volkswagen                       4015.0  
ford                             2621.0  
opel                             2264.0  
renault                          1785.0  
fiat                             1881.0  
smart                               NaN

We have aggregated mean brand prices per each mileage group. We can observe that the Volkswagen is definitely like a "buffer" between expensive and cheaper brands. We can ignore sonstige_autos, since that translates to other cars. Volkswagen also retains the price well (only 20% decrease) between low mileage and medium mileage groups, while Ford shows a price drop of 45% between for the same. 

## Exploring damage effect on the price

In [132]:
autos.columns

Index(['date_crawled', 'name', 'price_usd', 'ab_test', 'vehicle_type',
       'registration_year', 'gearbox', 'power_ps', 'model', 'odometer_km',
       'registration_month', 'fuel_type', 'brand', 'unrepaired_damage',
       'ad_created', 'postal_code', 'last_seen'],
      dtype='object')

In [134]:
no_damage = autos[autos["unrepaired_damage"] == "no"]
damage = autos[autos["unrepaired_damage"] == "yes"]

In [135]:
print(no_damage["price_usd"].mean())
print(damage["price_usd"].mean())

7165.752228057015
2282.2638540478906


We can see that, as expected, cars which have repaired damages are priced much higher than cars which have damages left unrepaired.


But, let's take a look which brands are more or less affected by unrepaired damage:


In [136]:
brands_unrepaired_vc = damage["brand"].value_counts(normalize = True).sort_values(ascending = False).head(10)
brands_unrepaired = brands_unrepaired_vc.index
brands_unrepaired

Index(['volkswagen', 'opel', 'ford', 'bmw', 'mercedes_benz', 'audi', 'renault',
       'peugeot', 'fiat', 'nissan'],
      dtype='object')

In [137]:
unrepaired_brand_price = {}

for b in brands_unrepaired:
    selected_rows = damage[damage["brand"] == b]
    mean_price = selected_rows["price_usd"].mean()
    unrepaired_brand_price[b] = int(mean_price)

In [138]:
unrepaired_brand_price

{'volkswagen': 2209,
 'opel': 1373,
 'ford': 1408,
 'bmw': 3560,
 'mercedes_benz': 4019,
 'audi': 3359,
 'renault': 1183,
 'peugeot': 1375,
 'fiat': 1182,
 'nissan': 1962}

In [139]:
ubp_series = pd.Series(unrepaired_brand_price).sort_values(ascending = False)

In [140]:
brands_repaired_vc = no_damage["brand"].value_counts(normalize = True).sort_values(ascending = False).head(10)
brands_repaired = brands_unrepaired_vc.index
brands_repaired

Index(['volkswagen', 'opel', 'ford', 'bmw', 'mercedes_benz', 'audi', 'renault',
       'peugeot', 'fiat', 'nissan'],
      dtype='object')

In [141]:
repaired_brand_price = {}

for b in brands_repaired:
    selected_rows = no_damage[no_damage["brand"] == b]
    mean_price = selected_rows["price_usd"].mean()
    repaired_brand_price[b] = int(mean_price)

In [142]:
rbp_series = pd.Series(repaired_brand_price).sort_values(ascending = False)

In [143]:
damage_price_info = pd.DataFrame(ubp_series, columns = ["unrepaired_price"])

In [144]:
damage_price_info["repaired_price"] = rbp_series

In [162]:
damage_price_info["diff"] = (((damage_price_info["unrepaired_price"] - damage_price_info["repaired_price"]) / damage_price_info["repaired_price"]) * 100).round()

In [166]:
damage_price_info.sort_values(by = ["diff"])

unrepaired_price  repaired_price  diff
ford                       1408            4697 -70.0
audi                       3359           10867 -69.0
volkswagen                 2209            6509 -66.0
nissan                     1962            5829 -66.0
fiat                       1182            3456 -66.0
peugeot                    1375            3697 -63.0
opel                       1373            3667 -63.0
bmw                        3560            9474 -62.0
renault                    1183            3082 -62.0
mercedes_benz              4019            9841 -59.0

In [164]:
damage_price_info["diff"].mean()

-64.6

We have aggregated Top 10 brands for repaired and unrepaired category and calculated their mean prices. Then, we have added another column showing the repaired model price increase compared to its unrepaired state.

We can observe that prices are 65% lower on average when cars have been unrepaired, compared to the cars which have repaired damage. The highest price drop can be seen with Ford and Audi, whose prices are 70% lower when left unrepaired, while Mercedes has the lowest price drop, but sells for ~12% higher price when unrepaired when compared to other high-end brands.

## Conclusion

We have cleaned up all the data, translated German to English to ensure better clarity, converted camlcase to snakecase and finally, ran some analyses. 

We have seen that prices are highly correlated to brands - high-end brands, naturally, cost more, but tend to retain the price better than some chaper brands. They also tend to have higher mileage since these brands are often used for long-range travel more frequently than smaller, compact city vehicles. Despite the high prices, these brands are very popular on the German market. Since these brands are of German origin, it may point to the fact that German population has a high level of trust towards German brands and are ready to spend more to get a brand they trust.

Volkswagen has shown to be the most popular brand with 21% of the market share. When we dug deeper, we saw that is due to the fact Volkswagens are right in the middle between high-end brands and chaper brands. They offer good quality for the price, a good bang for the buck so to say. They also tend to retain the price quite well with higher mileage.

Opel also has quite high market share, and lower average prices, which results in high popularity.

We have also analysed how the damage affect average prices. As expected, generally speaking, vehicles with unrepaired damages are selling for almost 70% lower prices. When we aggregate this by brands, we can see that Audis and Fords lose their value the most when left unrepaired, while Mercedes shows the least price drop.

Hope you enjoyed reading through this Project and that we have all managed to learn more about the German used-car market. Going further, we could analyse if higher HP vehicles are selling for more, but, since there is a lot of data with unrealistic HP, maybe this dataset is not the best for that purpose.